heim-process/src/sys/macos/bindings/process.rs

```diff
@@ -202,18 +208,25 @@ pub fn processes() -> Result<Vec<kinfo_proc>, Error> {
                 0,
             )
         };
-        match result {
-            libc::ENOMEM => continue,
-            code if code < 0 => return Err(Error::last_os_error().with_sysctl(name.as_ref())),
-            _ => {
-                let length = size / mem::size_of::<kinfo_proc>();
-                unsafe {
-                    processes.set_len(length);
-                }
-                debug_assert!(!processes.is_empty());
-
-                return Ok(processes);
+
+        if result < 0 {
+            // `libc::ENOMEM` indicates there was not enough space in `processes` to store the whole
+            // process list which can occur when a new process spawns between getting the size and
+            // storing. If this is the case then simply try again.
+            if io::Error::last_os_error().raw_os_error().unwrap() == libc::ENOMEM {
+                continue;
+            } else {
+                return Err(Error::last_os_error().with_sysctl(name.as_ref()));
             }
+        } else {
+            // Getting the list succeeded so let `processes` know how many processes it holds
+            let length = size / mem::size_of::<kinfo_proc>();
```

Comment:
As there is a `num_processes` defined above, this variable is redundant now

Reply:
This is actually required since the size may change between getting the initial reading used for `num_processes` and writing the values to `processes`. This would happen if there are fewer processes when storing the values compared to getting the initial size. If you think other people would be prone to be confused by this since the API for `sysctl` makes it easy to make these mistakes then I can try to give them two obviously separate names like `estimated_length` and `actual_length` or maybe add a comment?

Reply:
Ah, I see, it is my mistake of not reading the whole function body. That is fine to leave it as is, sorry for confusion :)